src/server/raft_client.rs

```diff
@@ -243,6 +242,22 @@ impl<T: RaftStoreRouter> RaftClient<T> {
     }
 }
 
+// Collect raft messages into a vector so that we can merge them into one message later.
+// `MAX_GRPC_SEND_MSG_LEN` will be considered when collecting.
+struct RaftMsgCollector(usize);
+impl BatchCollector<Vec<RaftMessage>, RaftMessage> for RaftMsgCollector {
+    fn collect(&mut self, v: &mut Vec<RaftMessage>, e: RaftMessage) -> Option<RaftMessage> {
+        let msg_size = e.compute_size() as usize;
+        if self.0 > 0 && self.0 + msg_size >= MAX_GRPC_SEND_MSG_LEN as usize {
```

Comment:
I think you need to leave some buffer for Batch message itself. Encoding of final messages should be larger than self.0 + msg_size.